// eslint-plugin-json-schema-validator/lib/configs/flat/base.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsoncParser = __importStar(require("jsonc-eslint-parser"));
const yamlParser = __importStar(require("yaml-eslint-parser"));
const tomlParser = __importStar(require("toml-eslint-parser"));
exports.default = [
    {
        plugins: {
            get "json-schema-validator"() {
                return require("../../index");
            },
        },
    },
    {
        files: [
            "*.json",
            "**/*.json",
            "*.json5",
            "**/*.json5",
            "*.jsonc",
            "**/*.jsonc",
        ],
        languageOptions: {
            parser: jsoncParser,
        },
        rules: {
            strict: "off",
            "no-unused-expressions": "off",
            "no-unused-vars": "off",
        },
    },
    {
        files: ["*.yaml", "**/*.yaml", "*.yml", "**/*.yml"],
        languageOptions: {
            parser: yamlParser,
        },
        rules: {
            "no-irregular-whitespace": "off",
            "no-unused-vars": "off",
            "spaced-comment": "off",
        },
    },
    {
        files: ["*.toml", "**/*.toml"],
        languageOptions: {
            parser: tomlParser,
        },
        rules: {
            "no-irregular-whitespace": "off",
            "spaced-comment": "off",
        },
    },
];
